sympy/polys/domains/domain.py

```diff
@@ -14,16 +14,301 @@
+    to a domain element and vice versa::
+
+        >>> from sympy import Symbol, ZZ, Expr
+        >>> x = Symbol('x')
+        >>> K = ZZ[x]           # polynomial ring domain
+        >>> K
+        ZZ[x]
+        >>> type(K)             # class of the domain
+        <class 'sympy.polys.domains.polynomialring.PolynomialRing'>
+        >>> K.dtype             # class of the elements
+        <class 'sympy.polys.rings.PolyElement'>
+        >>> p_expr = x**2 + 1   # Expr
+        >>> p_expr
+        x**2 + 1
+        >>> type(p_expr)
+        <class 'sympy.core.add.Add'>
+        >>> isinstance(p_expr, Expr)
+        True
+        >>> p_domain = K.from_sympy(p_expr)
+        >>> p_domain            # domain element
+        x**2 + 1
+        >>> type(p_domain)
+        <class 'sympy.polys.rings.PolyElement'>
+        >>> K.to_sympy(p_domain) == p_expr
+        True
+
+    The :py:meth:`~.Domain.convert_from` method is used to convert domain
+    elements from one domain to another::
+
+        >>> from sympy import ZZ, QQ
+        >>> ez = ZZ(2)
+        >>> eq = QQ.convert_from(ez, ZZ)
+        >>> type(ez)
+        <class 'int'>
+        >>> type(eq)
+        <class 'sympy.polys.domains.pythonrational.PythonRational'>
+
+    Elements from different domains should not be mixed in arithmetic or other
+    operations: they should be converted to a common domain first.  The domain
+    method :py:meth:`~.Domain.unify` is used to find a domain that can
+    represent all the elements of two given domains::
+
+        >>> from sympy import ZZ, QQ, symbols
+        >>> x, y = symbols('x, y')
+        >>> ZZ.unify(QQ)
+        QQ
+        >>> ZZ[x].unify(QQ)
+        QQ[x]
+        >>> ZZ[x].unify(QQ[y])
+        QQ[x,y]
+
+    If a domain is a :py:class:`~.Ring` then is might have an associated
+    :py:class:`~.Field` and vice versa. The :py:meth:`~.Domain.get_field` and
+    :py:meth:`~.Domain.get_ring` methods will find or create the associated
+    domain::
+
+        >>> from sympy import ZZ, QQ, Symbol
+        >>> x = Symbol('x')
+        >>> ZZ.has_assoc_Field
+        True
+        >>> ZZ.get_field()
+        QQ
+        >>> QQ.has_assoc_Ring
+        True
+        >>> QQ.get_ring()
+        ZZ
+        >>> K = QQ[x]
+        >>> K
+        QQ[x]
+        >>> K.get_field()
+        QQ(x)
+
+    See also
+    ========
+
+    DomainElement: abstract base class for domain elements
+
+    """
+
+    #: The type (class) of the elements of this :py:class:`~.Domain`:
+    #:
+    #: >>> from sympy import ZZ, QQ, Symbol
+    #: >>> ZZ.dtype
+    #: <class 'int'>
+    #: >>> z = ZZ(2)
+    #: >>> z
+    #: 2
+    #: >>> type(z)
+    #: <class 'int'>
+    #: >>> type(z) == ZZ.dtype
+    #: True
+    #:
+    #: Every domain has an associated **dtype** ("datatype") which is the
+    #: class of the associated domain elements.
+    #:
+    #: See also
+    #: ========
+    #:
+    #: of_type
+    #:
+    dtype = None        # type: Optional[Type]
```

Comment:
I like this style of documentation fo class attributes because it shows in the code and the docs but I don't think it's properly doctested. Is there a better/recommended way to document class attributes?